pandas/_testing/_warnings.py

```diff
@@ -32,7 +32,8 @@ def assert_produces_warning(
     ] = "always",
     check_stacklevel: bool = True,
     raise_on_extra_warnings: bool = True,
-    match: str | None = None,
+    match: str | tuple[str | None, ...] | None = None,
+    must_find_all_warnings: bool = False,
```

Comment:
I was just worried, because when I looked for its usages, IntelliJ found thousands of entries. I didn't want to break any tests that might depend on the previous behavior. The old behavior might also still prove to be useful sometimes.

Reply:
> I didn't want to break any tests that might depend on the previous behavior.

I would describe those tests as already broken and in need of fixing! I would hope that there aren't many tests that need changed and they can be updated here, but if that's not the case (let's say there are more than 30, not counting parametrizations), then that is definitely a reason to introduce it as you've done here and fix them here.

> The old behavior might also still prove to be useful sometimes.

I was thinking about this. The only case I can imagine where you'd have to use the old behavior is when you can't tell which warning are going to be emitted in an operation. That should be avoided if at all possible, and if it's not we could ignore the warnings.

Are there other cases where it might be useful?

Reply:
Agreed I think this this function should match all provided warnings & warning by default

Reply:
@rhshadrach With the new behavior set as a default one, we are getting 12 failed tests which presumably need fixing. For example, `pandas/tests/io/parser/common/test_inf.py::test_read_csv_with_use_inf_as_na[pyarrow]` fails as it doesn't find the `DeprecationWarning` and it really isn't there. What to do in that situation? Eliminate the `DeprecationWarning` from the expected warnings for that test?

Reply:
Also, all failed tests are about the `read_csv`, which doesn't seem to ever produce a `DeprecationWarning`. It seems that the tests were written, so they wouldn't break when the `FutureWarning` were to be replaced by the `DeprecationWarning`. That's why I believe it would be safe to exclude it from the test.

Reply:
Yep - removing the DeprecationWarning sounds great.

Reply:
Okay, I removed Warning where they aren't actually raised and also changed one util to correctly use the new API.